tensorflow/python/ops/special_math_ops.py

```diff
@@ -272,38 +289,64 @@ def einsum(equation, *inputs, **kwargs):
             ' because index "%s" is summed over more than two inputs.', a)
         return _exponential_space_einsum(equation, *inputs)
 
-    # Use xla_einsum if executing on TPU and if the operation is a 2 input
-    # einsum supported by XlaEinsumOp.
     if _enclosing_tpu_context() is not None and len(inputs) == 2:
       return gen_xla_ops.xla_einsum(
           inputs[0], inputs[1], input_axis_labels[0] + ',' +
           input_axis_labels[1] + '->' + output_axis_labels)
-    temp = inputs[0]
-    temp_axis_labels = input_axis_labels[0]
-    for i in xrange(len(inputs) - 1):
-      axes_to_sum = (
-          set(temp_axis_labels) &
-          set(input_axis_labels[i + 1]) - set(output_axis_labels))
-      temp, temp_axis_labels = _einsum_reduction(
-          temp, temp_axis_labels, inputs[i + 1], input_axis_labels[i + 1],
-          axes_to_sum)
-
-
-    missing_indices = set(temp_axis_labels) - set(output_axis_labels)
+
+    # the list seq gives the order of pairwise contractions; seq[0] is the
+    # first, seq[1] the second contraction and so forth; each element of seq
+    # (j,k) tells us to contract tensors j and k in the list of inputs, remove
+    # them from inputs and prepend the new tensor to the list
+    seq = einsum_optimize(
+      [t.shape.as_list() for t in inputs],
+      input_axis_labels, 
+      output_axis_labels,
+      optimize=optimize
+    )
+    
+    for j, k in seq:
+      if j > k: j, k = k, j
+      t2, l2 = inputs.pop(k), input_axis_labels.pop(k)
+      t1, l1 = inputs.pop(j), input_axis_labels.pop(j)
+      axes_to_sum = (set(l1) & set(l2)) - set(output_axis_labels)
+      t3, l3 = _einsum_reduction(t1, l1, t2, l2, axes_to_sum)
+      inputs.insert(0, t3)
+      input_axis_labels.insert(0, l3)
+    
+    missing_indices = set(input_axis_labels[0]) - set(output_axis_labels)
+
     if missing_indices:
       axis = [
-          i for i, a in enumerate(temp_axis_labels)
+          i for i, a in enumerate(input_axis_labels[0])
           if a not in output_axis_labels
       ]
-      temp = math_ops.reduce_sum(temp, axis=axis)
+      temp = math_ops.reduce_sum(inputs[0], axis=axis)
```

Comment:
These `temp` variables are unused. I added a much more direct test for this. It fails at the "Invalid Equation" error below.

You know this code better than me, can you take a look?